[stdlib] Unicode 9 and Thread Local Storage

milseman opened the pull request:
TL;DR Add support for Unicode 9 grapheme breaking (i.e. emoji) by using ICU. Work around ICU efficiency issues with thread-local storage.

---

Introduce shims for using UBreakIterators from ICU. Also introduce
shims for using thread local storage via pthreads.
    
We will be relying on ICU and UBreakIterators for grapheme
breaking. But, UBreakIterators are very expensive to create,
especially for the way we do grapheme breaking, which is relatively
stateless. Thus, we will stash one or more into thread local storage
and reset it as needed.
    
---

Introduces a _ThreadLocalStorage struct to hold thread-local, but
global resources. Set it up to host a UBreakIterator and a cache key
for resetting text.
    
UBreakIterators are extremely expensive to create on the fly, so we
store one for each thread. UBreakIterators are also expensive to bind
to new text, so we cache the text it's currently bound to in order to
help avoid it.
    
The struct can be expanded with more functionality in the future, but
the standard library should only ever use a single key, and thus
everything should go on this struct. The _ThreadLocalStorage struct is
not meant to be copyable, creatable (by anyone else except the
once-per-thread initialize routine), and should accessed through the
pointers it provides.
    
Future immediate directions could include cashing multiple
UBreakIterators (e.g. avoid a text reset for mutual character
iteration patterns, etc).
    
Test added in test/stdlib/ThreadLocalStorage.swift.

---

Use UBreakIterators to perform grapheme breaking. This gives Unicode 9
grapheme breaking (e.g. family emoji) and provides a means to upgrade
to future versions. It also serves as a model for how to serve up
other advanced functionality in ICU to users.
    
This has tricky performance implications. Some things are faster and a
number of cases are slower. But, careful use of ICU can help mitigate
and amortize these costs. In conjunction with more early detection of
fast paths, overall grapheme breaking for the average user should be
much faster than in Swift 3.
    
NOTE: This is incomplete. It currently falls back on the legacy tries
for some bridged strings. There are many potential directions for a
general solution, but for now we'll be interatively adding support for
more and more special cases.

milseman committed:
[stdlib] Shims for UBreakIterator and thread local storage.

Introduce shims for using UBreakIterators from ICU. Also introduce
shims for using thread local storage via pthreads.

We will be relying on ICU and UBreakIterators for grapheme
breaking. But, UBreakIterators are very expensive to create,
especially for the way we do grapheme breaking, which is relatively
stateless. Thus, we will stash one or more into thread local storage
and reset it as needed.

Note: Currently, pthread_key_t is hard coded for a single platform
(Darwin), but I have a static_assert alongside directions on how to
adapt it to any future platforms who differ in key type.

milseman committed:
[stdlib] Introduce Thread Local Storage

Introduces a _ThreadLocalStorage struct to hold thread-local, but
global resources. Set it up to host a UBreakIterator and a cache key
for resetting text.

UBreakIterators are extremely expensive to create on the fly, so we
store one for each thread. UBreakIterators are also expensive to bind
to new text, so we cache the text it's currently bound to in order to
help avoid it.

The struct can be expanded with more functionality in the future, but
the standard library should only ever use a single key, and thus
everything should go on this struct. The _ThreadLocalStorage struct is
not meant to be copyable, creatable (by anyone else except the
once-per-thread initialize routine), and should accessed through the
pointers it provides.

Future immediate directions could include cashing multiple
UBreakIterators (e.g. avoid a text reset for mutual character
iteration patterns, etc).

Test added in test/stdlib/ThreadLocalStorage.swift.

milseman committed:
[stdlib] Unicode 9 here we come: use ICU for grapheme breaking

Use UBreakIterators to perform grapheme breaking. This gives Unicode 9
grapheme breaking (e.g. family emoji) and provides a means to upgrade
to future versions. It also serves as a model for how to serve up
other advanced functionality in ICU to users.

This has tricky performance implications. Some things are faster and a
number of cases are slower. But, careful use of ICU can help mitigate
and amortize these costs. In conjunction with more early detection of
fast paths, overall grapheme breaking for the average user should be
much faster than in Swift 3.

NOTE: This is incomplete. It currently falls back on the legacy tries
for some bridged strings. There are many potential directions for a
general solution, but for now we'll be interatively adding support for
more and more special cases.

milseman commented:
@swift-ci please test 

Not just for kicks, I'm genuinely interested

milseman commented on stdlib/public/core/CMakeLists.txt:
```diff
@@ -174,6 +176,7 @@ if(CMAKE_SYSTEM_NAME STREQUAL "Darwin")
   list(APPEND swift_core_link_flags "-all_load")
   list(APPEND swift_core_framework_depends Foundation)
   list(APPEND swift_core_framework_depends CoreFoundation)
+  list(APPEND swift_core_private_link_libraries -licucore)
```
As @compnerd mentioned, we can remove the "-l" part

airspeedswift commented:
@swift-ci please test

airspeedswift commented:
@swift-ci please smoke benchmark

airspeedswift commented:
I think those benchmarks are showing changes on master from other things that went in since, like the new utf8 decoding and the change to remove <, not this change. Needs rebasing and rebenchmarking.

milseman commented:
@swift-ci please test

milseman committed:
[stdlib] Add ICU.swift, to host ICU helper functionality

milseman commented:
Rewrote history a little to squash some things, and only pull in the tiny bits of ICU that we use.

milseman commented:
Added in more static protections for cross-platform differences and a cache coherency fix.

milseman commented:
@swift-ci please smoke test Linux platform

milseman commented:
@swift-ci please smoke benchmark

milseman commented:
I extracted some of those smoke benchmarks, and I'm seeing us being way faster with this PR compared to a couple weeks ago (before a lot of my other optimizations). I'm measuring that we are:
StringWithCString: 5x faster
CStringLongAscii: 2.5x faster
CStringLongNonAscii: 2x faster

than a couple weeks ago.

milseman commented:
```
/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/swift/lib/ClangImporter/ClangImporter.cpp:2647:41: error: too few arguments to function call, single argument 'Diags' was not specified
  return Impl.Invocation->getModuleHash();
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
```

looks like the builder's broken, as other PRs are hitting this. Will try again soon.

milseman commented:
@swift-ci please smoke test Linux platform

CodaFi commented:
Just for bookkeeping: This supersedes #3125, right?

milseman commented:
@CodaFi when it lands, yes

milseman commented:
Sweet, that looks better. I'll look into those.

milseman commented:
Linux platform defines a different pthread_key_t, as suspected. Will fix.

milseman committed:
[stdlib] Linux definition of pthread_key_t

milseman commented:
@swift-ci please smoke test Linux Platform

milseman commented:
SuperChars is a legit slowdown. Grapheme breaking performance is sometimes faster and sometimes slower than the tries if we can't hit our fast paths. 

Some of the delta is the same inefficient Character construction from a (presumably) valid CharcterView.Index, but not most.

@airspeedswift Any ideas? Are you ok with this?

airspeedswift commented:
I think the regression in SuperChars is fine, that benchmark isn't very representative of real usage.

milseman committed:
[stdlib] Add Unicode 9 grapheme break tests for flags, emoji, etc.

milseman commented:
@swift-ci please smoke test Linux Platform

milseman commented:
@swift-ci please test OS X Platform

milseman commented:
@swift-ci please smoke test Linux Platform

milseman commented:
@swift-ci please smoke test Linux Platform

milseman committed:
[stdlib] Disable Unicode 9 test on Linux, depends on ICU version

milseman committed:
[stdlib stubs] Fix up Linux build to know about pthread types.

Adds in Linux platform support for our pthread TLS. Replace usage of
PTHREAD_KEYS_MAX with a sentinel value, as it's tricky to define
cross-platform and was only lightly used inside sanity checks.

milseman commented:
@swift-ci please smoke test

milseman committed:
[stdlib] Disable legacy Unicode 8.0 regional break test.

Disables the portions of the regional break test file corresponding to
Unicode 8.0 vs 9.0 differences. Once the compiler and stdlib are in
sync with Unicode 9.0, and the legacy tries are dropped, we should
splat down newer versions of these test files.

milseman commented:
@swift-ci please test

milseman commented:
Just to be extra cautious, I'm going to wait for that OS X test to finish...
